[MRG + 1] Implement haversine metric in pairwise

xuewei4d opened the pull request:
Fixes #4453 #4452 , add haversine distrance for pairwise_distances and document it.

~~I did not find any test for `pairwise.py`, so I did not add any test cases.~~

xuewei4d committed:
Fix #4453, add haversine distrance for pairwise_distances

ragv commented:
You could add a test [here](https://github.com/scikit-learn/scikit-learn/blob/master/sklearn/metrics/tests/test_pairwise.py) :)

ragv commented on sklearn/metrics/pairwise.py:
```diff
@@ -426,6 +426,57 @@ def pairwise_distances_argmin(X, Y, axis=1, metric="euclidean",
                                          metric_kwargs)[0]
 
 
+def haversine_distances(X, Y=None, ):
```
`(X, Y=None, )` --> `(X, y=None)`

ragv replied:
Ah sorry keep the y uppercased...

`(X, Y=None, ) --> (X, Y=None)`

ragv commented on sklearn/metrics/pairwise.py:
```diff
@@ -426,6 +426,57 @@ def pairwise_distances_argmin(X, Y, axis=1, metric="euclidean",
 
+def haversine_distances(X, Y=None, ):
+    """
+    Compute haversine distance between samples in X and Y.
```
`"""Compute haversine distance between samples in X and Y.` :)

xuewei4d commented:
@ragv I looked into wrong test folders :( 

Besides, the testing in ```test_neighbors.py``` failed in line 815.
```python
        nbrs = cls(algorithm='ball_tree', metric='haversine')
        assert_raises(ValueError,
                      nbrs.predict,
                      X)
        assert_raises(ValueError,
                      ignore_warnings(nbrs.fit),
                      Xsparse, y)
```

Is there any reason that haversine metric cannot be used with ball_tree? @jakevdp ?

ragv commented on sklearn/metrics/pairwise.py:
```diff
@@ -426,6 +426,57 @@ def pairwise_distances_argmin(X, Y, axis=1, metric="euclidean",
+        with shape (n_samples_X, 2).
+
+    Y : array_like
+        with shape (n_samples_Y, 2).
```
shape information can go in the first line itself... More description, if available, can follow next :) Like [here](https://github.com/scikit-learn/scikit-learn/blob/master/sklearn/metrics/pairwise.py#L159).

ragv commented on sklearn/metrics/pairwise.py:
```diff
@@ -426,6 +426,57 @@ def pairwise_distances_argmin(X, Y, axis=1, metric="euclidean",
+    Returns
+    -------
+    distance matrix : array
+        An array with shape (n_samples_X, n_samples_Y).
```
same here...

ragv commented on sklearn/metrics/pairwise.py:
```diff
@@ -426,6 +426,57 @@ def pairwise_distances_argmin(X, Y, axis=1, metric="euclidean",
+    if Y is None:
+        Y = X
+    if (X.shape[1] != 2 or Y.shape[1] != 2):
+        raise ValueError("Haversine distance only valid in 2 dimensions")
```
you should use [`check_paired_array`](https://github.com/scikit-learn/scikit-learn/blob/master/sklearn/metrics/pairwise.py#L101) as done [here](https://github.com/scikit-learn/scikit-learn/blob/master/sklearn/metrics/pairwise.py#L192).

ragv commented:
> Besides, the testing in test_neighbors.py failed in line 815.

You haven't validated the data like done [here](https://github.com/scikit-learn/scikit-learn/blob/master/sklearn/metrics/pairwise.py#L192) for `euclidean_distances`. :)

xuewei4d commented:
@ragv I don't think ```check_pairwise_arrays``` will let it pass ```test_neighbors_badargs``` in ```neighbors/test_neighbors.py```. It looks like ```        nbrs = cls(algorithm='ball_tree', metric='haversine')
``` is a bad argument.

ragv commented:
I suggested validation, simply upon seeing bad args in the test title... sorry :p

`haversine` metric was introduced in this PR only right? It says `ValueError` [was not raised](https://travis-ci.org/scikit-learn/scikit-learn/jobs/56242858#L1323)...

I *think* since previously `haversine` was not implemented, there was a test to check if `ValueError` was raised... Now that we have `haversine` implemented I think you can go ahead and remove [L815-L821](https://github.com/scikit-learn/scikit-learn/blob/master/sklearn/neighbors/tests/test_neighbors.py#L815-L821).

xuewei4d commented:
Thanks for your comments, @ragv. I think we need someone else to take a look at [L815](https://github.com/scikit-learn/scikit-learn/blob/0117fb54fc39184ae5fec941aaa742c832af48d0/sklearn/neighbors/tests/test_neighbors.py#L815-L821)

ragv commented:
There is an [unrelated test failure](https://travis-ci.org/scikit-learn/scikit-learn/jobs/56248127#L101) in Python 3.4 alone... :-/ Others [seem to pass](https://travis-ci.org/scikit-learn/scikit-learn/builds/56248124)...

xuewei4d commented:
Any review on this PR @amueller ? Should I need to check the data is in the domain of cos, sin? Spicy will give a running time warning anyway.

ragv commented:
You mean scipy right? :p

ragv commented:
Also in your PR description change fix to fixes so that github can automatically close that issue when this PR gets merged :)

amueller commented:
Hey @xuewei4d. Sorry I was sick over the weekend and just catching up. It is fine that ``BallTree`` doesn't work with this metric.

amueller commented on sklearn/metrics/tests/test_pairwise.py:
```diff
@@ -42,37 +43,61 @@
     S2 = pairwise_distances(X_tuples, Y_tuples, metric="euclidean")
     assert_array_almost_equal(S, S2)
+
+    # Test haversine distnace
```
distance

amueller commented on sklearn/metrics/tests/test_pairwise.py:
```diff
@@ -42,37 +43,61 @@
+    # Test haversine distnace
+    # The data should be valid latitude and longitude
+    X = rng.random_sample((5, 2))
+    X[:, 0] = (X[:, 0] - 0.5)*2*np.pi/2
```
pep8

amueller commented on sklearn/neighbors/tests/test_neighbors.py:
```diff
@@ -812,13 +812,6 @@ def test_neighbors_badargs():
                       cls, p=-1)
         assert_raises(ValueError,
                       cls, algorithm='blah')
-        nbrs = cls(algorithm='ball_tree', metric='haversine')
```
This still crashes, right? Why did you remove the test?

amueller replied:
So you got "error not raised" here?

ragv replied:
Sorry this was my suggestion.

> So you got "error not raised" here?

Yes. Travis failed owing to this. I assumed since haversine was not previously implemented, this test was introduced.

amueller replied:
Yes, but I thought it is still not implemented for the ball_tree. Is it?

xuewei4d replied:
I think it is implemented [L285](https://github.com/scikit-learn/scikit-learn/blob/master/sklearn/neighbors/tests/test_ball_tree.py#L285) in ```test_ball_tree.py``` [L22](https://github.com/scikit-learn/scikit-learn/blob/master/sklearn/neighbors/ball_tree.pyx#L22) in ```ball_tree.pyx```. The test  is to make sure haversine is not implemented for four ```neighbors``` classifiers and regressors.

amueller replied:
Ah, ok. The test was slightly confusing to me.

amueller commented:
You should add the metric to ``test_neighbors_metrics`` in ``test_neighbors.py``

amueller commented:
Does the tree also raise an exception if the data is not 2d? Maybe that should be tested in test_neighbors_badargs?

amueller commented on sklearn/metrics/pairwise.py:
```diff
@@ -426,6 +426,53 @@ def pairwise_distances_argmin(X, Y, axis=1, metric="euclidean",
+    >>> Y = np.array([[-np.pi/2, 0], [-np.pi/2, -np.pi]])
+    >>> haversine_distances(X, Y) # doctest: +ELLIPSIS
+    array([[ 1.57...,  1.57...],
+           [ 3.14...,  3.14...]])
```
Can you add a reference (to wikipedia?)

amueller commented on sklearn/metrics/pairwise.py:
```diff
@@ -426,6 +426,53 @@ def pairwise_distances_argmin(X, Y, axis=1, metric="euclidean",
+
+    D = np.square(np.sin((X[:, np.newaxis, :] - Y[np.newaxis, :, :]) * 0.5))
+    D = 2 * np.arcsin(np.sqrt(D[:, :, 0] +
+                              np.dot(np.cos(X[:, np.newaxis, 0]),
```
maybe it's shorter to write ``np.outer(np.cos(X[:, 0]), np.cos(Y[:, 0]))``. That is what you want here, right?

xuewei4d replied:
Exactly! I didn't know ```np.outer```   :-|

amueller commented:
LGTM apart from my minor comments above.

xuewei4d commented:
> You should add the metric to test_neighbors_metrics in test_neighbors.py

I will do that.

BTW, I found [L871](https://github.com/scikit-learn/scikit-learn/blob/master/sklearn/neighbors/tests/test_neighbors.py#L871) test if ```metric``` is a valid metric for kd_tree, but does not test it for ball_tree. Is this an issue?

xuewei4d commented:
> Does the tree also raise an exception if the data is not 2d? Maybe that should be tested in test_neighbors_badargs?

ball_tree does raise an exception,  I will add code to raise an exception in ```NeighborsBase._fit```

amueller commented:
You shouldn't change previous behavior. What does ``ball_tree`` do with higher dimensions? The surface distance of n-dim balls?

amueller commented:
> BTW, I found L871 test if metric is a valid metric for kd_tree, but does not test it for ball_tree. Is this an issue?
I think the list of metrics that is tested is just those implemented in ``BallTree``.

xuewei4d commented:
> You shouldn't change previous behavior. What does ball_tree do with higher dimensions? The surface distance of n-dim balls?

I don't quite understand your point. I mean if we use the current master branch implementation, 
```python
nbrs = neighbors.NearestNeighbors(algorithm='ball_tree', metric='haversine')
X = np.random.rand(10, 3)
nbrs.fit(X)
```
will raise ```ValueError``` from 
```python
binary_tree.pxi in sklearn.neighbors.ball_tree.BinaryTree.__init__ (sklearn/neighbors/ball_tree.c:8825)()

binary_tree.pxi in sklearn.neighbors.ball_tree.BinaryTree._recursive_build (sklearn/neighbors/ball_tree.c:10057)()

ball_tree.pyx in sklearn.neighbors.ball_tree.init_node (sklearn/neighbors/ball_tree.c:19499)()

binary_tree.pxi in sklearn.neighbors.ball_tree.BinaryTree.rdist (sklearn/neighbors/ball_tree.c:9943)()

dist_metrics.pyx in sklearn.neighbors.dist_metrics.HaversineDistance.rdist (sklearn/neighbors/dist_metrics.c:8764)()
```
which is not very friendly. So I added checking code in ```NeighborsBase._fit``` which would be more better. Do you think I don't have to do this?

xuewei4d commented:
Let me recap the issue #4452. When DBSCAN with auto algorithm and haversine metric, it call ```pairwise_distances``` in pairwise.py, but haversine distance have not been implemented for ```pairwise_distances``` yet before this PR. I implemented it. DBSCAN with ball_tree and haversine works fine, since haversine is already there for ball_tree.

For  ```n_feature``` checking,  there is  already ```n_features``` check in dist_metrics.pyx. Therefore, ```neighbors.KNeighborsClassifier```, ```neighbors.KNeighborsRegressor```, ```neighbors.RadiusNeighborsClassifier``` and ```neighbors.RadiusNeighborsRegressor``` will raise an ValueError  in ```fit``` with ball_tree. It won't raise it in ```fit``` when using brute or auto algorithm, but raise it in ```predict```.  Same thing happens for ```neighbor.NearestNeighbor```, when it is with ```brute``` algorithm, it raises ValueError only in ```kneighbors()```, but with ```ball_tree``` it raises in ```fit``` by ```haversine_distances```I just implemented in ```pairwise.py```. For these inconsistent raising behaviors, I think it will be better to check ```n_features``` for haversine distance in the very early stage in all ```fit```, i.e.  ```NeighborsBase._fit```, then all these classes raises a ValueError  in ```fit``` if ```n_feature``` is not 2 when using haversine distance.

amueller commented:
Sorry, misunderstood your previous comment (to much task switching?) as that the trees work for arbitrary ``n_features``.  If they raise a value error, then the ``pairwise_distance`` implementation should do the same. For the "friendlyness" of the error, I would rather go by the message than by the traceback. If the message is comprehensible, I wouldn't worry about a long traceback. For erroring in ``fit`` vs in ``predict`` I am not too concerned. I would rather not bloat ``NeighborsBase._fit`` with metric-specific checking.

xuewei4d commented:
Thanks, your comment makes more sense. So back to the testing, I think, the additional testing for bad arguments would be 
all nearestneighbor-like classes with brute/auto algorithm and haversine distance on data that ```n_feature != 2``` right?

amueller commented:
yes, that sounds good.

amueller commented:
LGTM

amueller commented:
Are there any todos? Otherwise, rename to [MRG]

xuewei4d commented:
May I squash all these commits?

amueller commented:
Yes please.

xuewei4d commented:
Not sure about Atlas problem reported by Travis.... Needs restart?

amueller commented:
restarted.

jakevdp commented:
Yes, haversine distances are defined only in two dimensions. See https://github.com/scikit-learn/scikit-learn/blob/master/sklearn/neighbors/dist_metrics.pyx#L955

jakevdp commented:
Just out of curiosity, could you check the performance of this algorithm vs. using the ``DistanceMetric`` class directly? e.g.
``` python
In [1]: from sklearn.neighbors import DistanceMetric

In [2]: import numpy as np

In [3]: X = np.random.random((100, 2))

In [20]: metric = DistanceMetric.get_metric('haversine')

In [21]: dist = metric.pairwise(X, X)
```
The difference is that this doesn't do any broadcasting, so it may be faster and/or more memory efficient.

Note that the ``DistanceMetric`` stuff is not really documented in the package, as I considered it primarily a workhorse for the BallTree/KDTree. The only reason the Python interface exists is to facilitate unit tests, but there's no reason we can't take advantage of it elsewhere!

amueller commented:
I wasn't really aware of the DistanceMetric interface. Shame on me :-/ could we also use that for the jaccard?

jakevdp commented:
Yes, it would work for Jaccard! Actually, it will work for any of the metrics supported by BallTree. For pairwise distances, it tends to be a factor of a few slower than the equivalent scipy routines, though (if I remember correctly; I did the benchmarks a long time ago).

amueller commented:
Ok we should check for jaccard.

amueller commented:
what was the status here? Just waiting for another review?

xuewei4d commented:
I think I need to do some benchmark.

raghavrv commented:
@xuewei4d Would you be having time to finish this up?

ibrahimsharaf commented:
Hi all, How can I finish this?

jnothman commented:
Could you please benchmark how fast this is compared to an implementation based on Distance metric as suggested by jakevdp above?

jnothman commented:
Closing due to being superseded by #12568
